File: hooks/pin-input.ts
import * as pinInput from "@zag-js/pin-input";
import { normalizeProps, spreadProps, renderPart, getBooleanOption, getOption } from "./util";
import { Component } from "./component";
import type { ViewHook } from "phoenix_live_view";
import type { Machine } from "@zag-js/core";

type Type = "alphanumeric" | "numeric" | "alphabetic" | undefined;
type Dir = "ltr" | "rtl" | undefined;

class PinInput extends Component<pinInput.Context, pinInput.Api> {
  initService(context: pinInput.Context): Machine<any, any, any> {
    return pinInput.machine(context);
  }

  initApi() {
    return pinInput.connect(this.service.state, this.service.send, normalizeProps);
  }

  clearValue() {
    this.api.clearValue();
  }

  render() {
    const parts = ["root", "label"];
    for (const part of parts) renderPart(this.el, part, this.api);
    this.renderInputs();
  }

  renderInputs() {
    for (const input of this.el.querySelectorAll<HTMLElement>("[data-part='input']")) {
      const index = input.dataset.index;
      if (!index || Number.isNaN(Number.parseInt(index))) {
        console.error("Missing or non-integer `data-index` attribute on input.");
        return;
      }
      spreadProps(input, this.api.getInputProps({ index: Number.parseInt(index) }));
    }
  }
}

export interface PinInputHook extends ViewHook {
  pinInput: PinInput;
  context(): pinInput.Context;
}

export default {
  mounted() {
    this.pinInput = new PinInput(this.el, this.context());
    this.pinInput.init();

    this.handleEvent("clear", () => this.pinInput.clearValue());
  },

  updated() {
    this.pinInput.render();
  },

  beforeDestroy() {
    this.pinInput.destroy();
  },

  context(): pinInput.Context {
    return {
      id: this.el.id,
      placeholder: this.el.dataset.placeholder,
      type: getOption(this.el, "type", ["alphanumeric", "numeric", "alphabetic"] as const) as Type,
      dir: getOption(this.el, "dir", ["ltr", "rtl"] as const) as Dir,
      otp: getBooleanOption(this.el, "otp"),
      mask: getBooleanOption(this.el, "mask"),
      blurOnComplete: getBooleanOption(this.el, "blurOnComplete"),
      onValueChange: (details: pinInput.ValueChangeDetails) => {
        if (this.el.dataset.onChange) {
          this.pushEvent(this.el.dataset.onChange, details);
        }
      },
      onValueComplete: (details: pinInput.ValueChangeDetails) => {
        if (this.el.dataset.onComplete) {
          this.pushEvent(this.el.dataset.onComplete, details);
        }
      },
      onValueInvalid: (details: pinInput.ValueInvalidDetails) => {
        if (this.el.dataset.onInvalid) {
          this.pushEvent(this.el.dataset.onInvalid, details);
        }
      },
    };
  },
} as PinInputHook;

File: lib/turboprop/headless/pin_input.ex
defmodule Turboprop.Headless.PinInput do
  @moduledoc false
  use Phoenix.Component

  import Turboprop.Headless

  attr :id, :string, default: nil
  attr :as, :any, default: "div"

  attr :type, :string, values: ["alphanumeric", "numeric", "alphabetic"], default: "numeric"
  attr :placeholder, :string, default: nil
  attr :dir, :string, values: ["ltr", "rtl"], default: "ltr"
  attr :otp, :boolean, default: false
  attr :blur_on_complete, :boolean, default: true

  attr :on_change, :string, default: nil
  attr :on_complete, :string, default: nil
  attr :on_invalid, :string, default: nil

  attr :rest, :global
  slot :inner_block

  def accordion(assigns) do
    {type, assigns} = Map.pop(assigns, :type)
    {placeholder, assigns} = Map.pop(assigns, :placeholder)
    {dir, assigns} = Map.pop(assigns, :dir)
    {otp, assigns} = Map.pop(assigns, :otp)
    {blur_on_complete, assigns} = Map.pop(assigns, :blur_on_complete)

    {on_change, assigns} = Map.pop(assigns, :on_change)
    {on_complete, assigns} = Map.pop(assigns, :on_complete)
    {on_invalid, assigns} = Map.pop(assigns, :on_invalid)

    render_as_tag_or_component(assigns, %{
      "id" => assigns.id || id(),
      "phx-hook" => "PinInput",
      "data-type" => type,
      "data-placeholder" => placeholder,
      "data-dir" => dir,
      "data-otp" => otp,
      "data-blur-on-complete" => blur_on_complete,
      "data-on-change" => on_change,
      "data-on-complete" => on_complete,
      "data-on-invalid" => on_invalid
    })
  end

  attr :as, :any, default: "div"
  attr :rest, :global
  slot :inner_block

  def root(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "root"})
  end

  attr :as, :any, default: "input"
  attr :index, :integer, required: true
  attr :rest, :global
  slot :inner_block

  def input(assigns) do
    {index, assigns} = Map.pop(assigns, :index)

    render_as_tag_or_component(assigns, %{"data-part" => "input", "data-index" => index})
  end

  attr :as, :any, default: "span"
  attr :rest, :global
  slot :inner_block

  def label(assigns) do
    render_as_tag_or_component(assigns, %{"data-part" => "label"})
  end
end
